include/openmc/cell.h

```diff
@@ -286,10 +288,40 @@ class UniversePartitioner
   std::vector<std::vector<int32_t>> partitions_;
 };
 
+//==============================================================================
+//! Define an instance of a particular cell
+//==============================================================================
+
+struct CellInstance {
+  //! Check for equality
+  bool operator==(const CellInstance& other) const
+  { return index_cell == other.index_cell && instance == other.instance; }
+
+  gsl::index index_cell;
+  gsl::index instance;
+};
+} // namespace openmc
+
+namespace std {
+template<>
+struct hash<openmc::CellInstance> {
+  // Taken from https://stackoverflow.com/a/17017281
+  std::size_t operator()(const openmc::CellInstance& k) const
+  {
+    std::size_t res = 17;
+    res = 31 * res + std::hash<gsl::index>()(k.index_cell);
+    res = 31 * res + std::hash<gsl::index>()(k.instance);
+    return res;
+  }
+};
+} // namespace std
```

Comment:
This hash function doesn't work very well for our purposes. With GCC, the `hash` of an integer just returns the same integer, so this function boils down to `16337 + 31*index_cell + instance`

That means that `hash({1, 0}) = hash({0, 31})` and `hash({1, 1}) = hash({0, 32})` and so on. We probably want a hash function with a period roughly larger than what we expect for typical distribcell sizes. And I don't think that added constant does anything at all for us. Playing around with the hash function a bit, it looks like this really cheap function gives decent results:
```c++
    return 0x1000*k.index_cell + k.instance;
```

If you want to see for yourself, here is a simple piece of test code that helps for testing different hash functions:
```c++
extern "C" void test_map() {
  std::unordered_map<CellInstance, std::int64_t> map;

  // Populate an example map.
  for (int cell = 0; cell < 10; ++cell) {
    for (int inst = 0; inst < 1000; ++inst) {
      CellInstance key {cell, inst};
      map[key] = 0;
    }
  }

  // Print the number of elements in each bucket.
  for (int i = 0; i < map.bucket_count(); ++i) {
    std::cout << map.bucket_size(i) << " ";
  }
  std::cout << "\n";
}
```

Reply:
Oh also, maybe we should pass the hash function in as a template argument to `unordered_map`. That feels cleaner than modifying the `std` namespace.

Reply:
Wow, that hash really does suck for two ints. Thanks for looking into that closely! I'll adopt your very simple hash function and move that into the template argument :+1: